politeiad/backendv2/tstorebe/tstorebe.go

```diff
@@ -0,0 +1,1251 @@
+	// record so that it can perform multiple read/write operations
+	// in a concurrent safe manner. These mutexes are lazy loaded.
+	recordMtxs map[string]*sync.Mutex
+}
+
+// isShutdown returns whether the backend is shutdown.
+func (t *tstoreBackend) isShutdown() bool {
+	t.RLock()
+	defer t.RUnlock()
+
+	return t.shutdown
+}
+
+// recordMutex returns the mutex for a record.
+func (t *tstoreBackend) recordMutex(token []byte) *sync.Mutex {
+	t.Lock()
+	defer t.Unlock()
+
+	ts := hex.EncodeToString(token)
+	m, ok := t.recordMtxs[ts]
+	if !ok {
+		// recordMtxs is lazy loaded
+		m = &sync.Mutex{}
+		t.recordMtxs[ts] = m
+	}
+
+	return m
+}
+
+// tokenCollision returns whether the short version of the provided token
+// already exists. This can be used to prevent collisions when creating new
+// tokens.
+func (t *tstoreBackend) tokenCollision(fullToken []byte) bool {
+	shortToken, err := util.ShortTokenEncode(fullToken)
+	if err != nil {
+		return false
+	}
+
+	t.RLock()
+	defer t.RUnlock()
+
+	_, ok := t.tokens[shortToken]
+	return ok
+}
+
+// tokenAdd adds a entry to the tokens cache.
+func (t *tstoreBackend) tokenAdd(fullToken []byte) error {
+	if !tokenIsFullLength(fullToken) {
+		return fmt.Errorf("token is not full length")
+	}
+
+	shortToken, err := util.ShortTokenEncode(fullToken)
+	if err != nil {
+		return err
+	}
+
+	t.Lock()
+	t.tokens[shortToken] = fullToken
+	t.Unlock()
+
+	log.Tracef("Token cache add: %v", shortToken)
+
+	return nil
+}
+
+// fullLengthToken returns the full length token given the short token. A
+// ErrRecordNotFound error is returned if a record does not exist for the
+// provided token.
+func (t *tstoreBackend) fullLengthToken(token []byte) ([]byte, error) {
+	if tokenIsFullLength(token) {
+		// Token is already full length. Nothing else to do.
+		return token, nil
+	}
+
+	shortToken, err := util.ShortTokenEncode(token)
+	if err != nil {
+		// Token was not large enough to be a short token. This cannot
+		// be used to lookup a record.
+		return nil, backend.ErrRecordNotFound
+	}
+
+	t.RLock()
+	defer t.RUnlock()
+
+	fullToken, ok := t.tokens[shortToken]
+	if !ok {
+		// Short token does not correspond to a record token
+		return nil, backend.ErrRecordNotFound
+	}
+
+	return fullToken, nil
+}
+
+func tokenIsFullLength(token []byte) bool {
+	return util.TokenIsFullLength(util.TokenTypeTstore, token)
+}
+
+func tokenFromTreeID(treeID int64) []byte {
+	b := make([]byte, 8)
+	// Converting between int64 and uint64 doesn't change
+	// the sign bit, only the way it's interpreted.
+	binary.LittleEndian.PutUint64(b, uint64(treeID))
+	return b
+}
+
+func treeIDFromToken(token []byte) int64 {
+	if !tokenIsFullLength(token) {
+		return 0
+	}
+	return int64(binary.LittleEndian.Uint64(token))
+}
+
+// metadataStreamsVerify verifies that all provided metadata streams are sane.
+func metadataStreamsVerify(metadata []backend.MetadataStream) error {
+	// Verify metadata
```

Comment:
Is this duplicate code? I think i have seen this before, including the next function.

Reply:
No, this validation is specific to politeiad.  Variations of this function exist in the plugins, but that validation is specific to the plugin.